### PR TITLE
Edit post

### LINKED ISSUE #73: [contracts] Edit post functionality
We can allow a user to edit a post by providing a proof of epoch key ownership, the old hash of the post, and the new content.

Includes a function taking an old hash + new hash + proof.

### LINKED ISSUE #80: [backend] Edit post functionality
- Accept a proof, send the transaction
- Hash the post content
- Update the database

### PR DESCRIPTION
**Pull request by vivianjeng**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

- closes #73 
- closes #80 
-

## Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

**Comment by vivianjeng** [comment]
Since we need post id or comment id to edit on-chain hashed content, so I proposed `contentId` as the post or comment id.
I think we cannot use transaction hash in `publishPost` because solidity doesn't support to get current transaction hash.
So `contentId` will be used to edit posts/comment on-chain
`transactionHash` will be used in frontend to query post/comment or link to etherscan
(Since the function will not return `contentId` but it will return `transactionHash`)
and `_id` is automatically generated by db (or it can be either `contentId` or `transactionHash`?)

**Comment by kittybest** [comment]
not sure if I understand you correctly -- if you're saying that the `postId` and the `commentId` should not be as same as `txHash`, actually now they are really not the same (as I saw in the `main` branch`), so actually you don't have to retrieve `postId` or `commentId` by `txHash` in the backend.

File: packages/core/contracts/UnirepSocial.sol [hunk 1]
```diff
@@ -70,19 +76,26 @@ contract UnirepSocial is zkSNARKHelper {
...
         uint256 indexed _epochKey,
-        string _commentContent,
+        uint256 _commentId,
+        bytes32 _contentHash,
         uint256 minRep
     );
 
+    event ContentUpdated(
+        uint256 indexed _id,
+        bytes32 _newContent
```
Comment (vimwitch) [inline packages/core/contracts/UnirepSocial.sol:96]: Might be good to emit and `oldContentHash` field

**Review by vimwitch — COMMENTED** [review]
We have some weird stuff around ids now. Posts and comments have a default id generated when they are created (`nanoid()`). I like the approach of giving an id based on an integer stored on chain, but it causes problems around post creation. It's impossible to know the id until _after_ the transaction has been processed. We just need to design around that.

I think we can use the current `_id` field and have the onchain one as well. The onchain one survives resyncs.

**Comment by vivianjeng** [comment]
Now this PR doesn't break the frontend & post deletion feature is added
Now `edit` and `delete` is the same function in the contract
`delete` is to set the `newHashContent = 0` so it just changes the backend.

`nanoid` is used in frontend and to query the posts/comments
`contentId` is used in contract to edit the content
so user won't need to know the `contentId` if they use unirep social frontend
all comments/votes are sent to specific `nanoid` (either post or comment)
though we cannot predict the `contentId` but we only need this variable once it is a valid transaction (`status=1`)

@kittybest @vimwitch Please check

File: packages/backend/src/routes/CommentRouter.ts [hunk 2]
```diff
@@ -246,3 +254,142 @@ async function createCommentSubsidy(req, res) {
...
+    const comment = await req.db.findOne('Comment', { where: { _id: id } })
+
+    res.json({
+        error: error,
+        transaction: hash,
+        comment,
+    })
+}
+
+async function deleteComment(req, res) {
```
Comment (vimwitch) [inline packages/backend/src/routes/CommentRouter.ts:329]: We're not actually deleting the comment here, which makes sense. Instead of having a custom backend route why don't we have the frontend update the comment content to `Deleted by user.` or similar. This delete function is a bit misleading

File: packages/backend/src/routes/PostRouter.ts [hunk 3]
```diff
@@ -233,3 +233,143 @@ async function createPostSubsidy(req, res) {
...
+    const post = await req.db.findOne('Post', { where: { _id: id } })
+
+    res.json({
+        error: error,
+        transaction: hash,
+        post,
+    })
+}
+
+async function deletePost(req, res) {
```
Comment (vimwitch) [inline packages/backend/src/routes/PostRouter.ts:308]: Same here, on the frontend let's just have it update the post content to `Deleted by user` and remove this route.

File: packages/backend/src/routes/CommentRouter.ts [hunk 4]
```diff
@@ -246,3 +254,142 @@ async function createCommentSubsidy(req, res) {
...
+    const { publicSignals, proof, content } = req.body
+    const epkProof = new EpochKeyProof(
+        publicSignals,
+        formatProofForSnarkjsVerification(proof)
+    )
+    const newHashedContent = ethers.utils.keccak256(
+        ethers.utils.toUtf8Bytes(content)
+    )
+
+    const error = await verifyEpochKeyProof(req.db, epkProof)
```
Comment (vimwitch) [inline packages/backend/src/routes/CommentRouter.ts:276]: We need to check that the `epoch` of the proof is the same epoch that the comment was created. We already store this on the object.

File: packages/core/contracts/UnirepSocial.sol [hunk 5]
```diff
@@ -368,29 +410,65 @@ contract UnirepSocial is zkSNARKHelper {
...
+     * @param oldContentHash The old hashed content of the post/comment
+     * @param newContentHash The new hashed content of the post/comment
+     * @param publicSignals The public signals of the epoch key proof of the author of the post/comment
+     * @param proof The epoch key proof of the author of the post/comment
+     */
+    function edit(
+        uint256 id,
+        bytes32 oldContentHash,
+        bytes32 newContentHash,
+        uint256[] memory publicSignals,
```
Comment (vimwitch) [inline packages/core/contracts/UnirepSocial.sol:460]: We really need a signing system for epoch keys, like semaphore has. The user should prove their epoch key and the `id`/`oldContentHash`/`newContentHash`. Then the backend can't modify the content.

We can do this in a different pr though.

File: packages/core/src/schema.ts [hunk 6]
```diff
@@ -33,6 +33,7 @@ const _schema = [
                 default: () => +new Date(),
             },
             ['postId', 'String'],
+            ['commentId', 'String', { optional: true }],
```
Comment (vimwitch) [inline packages/core/src/schema.ts:36]: ah this is confusing. A post has `_id` and `postId` and a comment has a `_id` and a `commentId`. Instead can we call this `onchainId` or something like that for both?

File: packages/core/test/Posting.ts [hunk 7]
```diff
@@ -269,15 +269,16 @@ describe('Post', function () {
                 const hashedContent = ethers.utils.keccak256(
                     ethers.utils.toUtf8Bytes(content)
                 )
-                const receipt = await unirepSocialContract
+                const { logs } = await unirepSocialContract
```
Comment (vimwitch) [inline packages/core/test/Posting.ts:272]: You can make an assertion about the events emitted by the contract using hardhat/chai. It's much cleaner to use [this](https://github.com/Unirep/Unirep/blob/v1.1/packages/contracts/test/epochTransition.test.ts#L565) kind of syntax i think (also supports data params)

**Review by vimwitch — CHANGES_REQUESTED** [review]
Looks good, just need to remove some redundant code and rename a couple things

File: packages/core/test/Posting.ts (hunk 7 above)
Comment (vivianjeng) [reply packages/core/test/Posting.ts:272]: I think here we only need `postId` but we don't have to make assertion.
So I parse the logs here to get `postId`

Comment (vimwitch) [reply packages/core/test/Posting.ts:272]: Ah sure, that makes sense.

File: packages/backend/src/routes/CommentRouter.ts [hunk 8]
```diff
@@ -149,6 +148,7 @@ async function createComment(req, res) {
 
     const comment = await req.db.create('Comment', {
         postId,
+        onChainPostId: post.onChainPostId,
```
Comment (vimwitch) [inline packages/backend/src/routes/CommentRouter.ts:151]: What do you think of removing `onChainPostId` and having `postId` be that value?

It shouldn't be possible to make a comment on a post until the post exists onchain, right?

File: packages/core/src/schema.ts [hunk 9]
```diff
@@ -32,8 +32,9 @@ const _schema = [
                 type: 'Int',
                 default: () => +new Date(),
             },
-            ['postId', 'String'],
-            ['commentId', 'String', { optional: true }],
+            ['postId', 'String', { optional: true }],
+            ['onChainPostId', 'String'],
+            ['onChainCommentId', 'String', { optional: true }],
```
Comment (vimwitch) [inline packages/core/src/schema.ts:37]: This can just be `onChainId`. If we're referencing this on a comment it's like `comment.onChainId` instead of `comment.onChainCommentId`.

File: packages/core/src/schema.ts [hunk 10]
```diff
@@ -67,7 +68,7 @@ const _schema = [
                 type: 'Int',
                 default: () => +new Date(),
             },
-            ['postId', 'String', { optional: true }],
+            ['onChainPostId', 'String', { optional: true }],
```
Comment (vimwitch) [inline packages/core/src/schema.ts:71]: Same here, just `onChainId` should be more clear.

**Review by vimwitch — CHANGES_REQUESTED** [review]

**Review by vimwitch — APPROVED** [review]

**Review by kittybest — APPROVED** [review]
